nnf/tseitin.py

```diff
@@ -0,0 +1,47 @@
+"""Transformations using the well-known Tseitin encoding <https://en.wikipedia.org/wiki/Tseytin_transformation>
+
+The Tseitin transformation converts any arbitrary circuit to one in CNF in
+polynomial time/space. It does so at the cost of introducing new variables
+(one for each logical connective in the formula).
+"""
+
+from nnf import NNF, Var, And, Or, Name, true, false
+
+def to_cnf(theory: NNF) -> NNF:
+
+    node_to_var = {}
+    clauses = []
+
+    def process_node(node: NNF) -> NNF:
+
+        if node in node_to_var:
+            return node
+
+        if isinstance(node, Var):
+            node_to_var[node] = node
+            return node
+
+        aux = Var("aux_%d" % len(node_to_var))
```

Comment:
Good catch! I think that makes loads of sense. Couldn't it just subclass from `Var` and instead of `v = Var('aux_0')`, we'd have `v = Aux()` or `v = AuxVar()`?

Global counter makes sense to me, and for (un/)pickling, we could overload the appropriate operators to re-encode the aux var labels (and set the global counter appropriately). It would violate some invariants, such as pickling an unpickled theory (after another is already loaded) creates something different than the original.

Actually, thinking through it further, if I've stored multiple theories, I may want to load them and have their variables refer to the same thing -- potentially the aux ones as well (e.g., encode something, compile to d-DNNF, condition in multiple ways, store, load and combine later). So perhaps the simpler option (no renaming or bothering on load) for now, and a future feature to change a config and rename on load (and potentially even save).

Reply:
Hm, subclassing from `Var` doesn't seem worth the cost to boringness and predictability. It's nice to be certain that there are only three types that can be instantiated. An interface to get the var immediately would be nice though. Could be `Var.aux()`, `Aux.var()`, `aux()`, `auxvar()`.

I'm tending more toward UUIDs now. If the only problem is that they look ugly then I'd rather do the display in a clever fiddly way than the safety.

To make the labels less ugly we could show only a few (4-6?) characters of the UUID. That'll be enough for most visual inspection.

Another option could be to keep a counter just for the display, with a ([weakref?](https://docs.python.org/3/library/weakref.html#weakref.WeakKeyDictionary)) mapping from UUIDs to ints. But that would be more complicated and inconsistent from run to run.

Reply:
Hrmz...preferences would be:
- `Var.aux()` (should store something locally in the var so we can identify it easily as an aux var)
- UUIDs, and 4-char display

We don't really gain anything by having indices in sequence from an arbitrary creation order.

Reply:
Added in dbdac4a03a443d55cfc4c7254eb312df5abf06e9. For now I've made `Aux` a trivial subclass of `UUID`, so you can do `isinstance(var.name, Aux)`.